File: QR-code/source code/Cpp/src/QR.cpp
#include "QR.h"
#include <cmath>

#define NUMERICO 1
#define ALPHANUMERICO 2
#define BYTE 3

#define  L 1
#define  M 2
#define  Q 3
#define  H 4

enum {
    Msg_n_suport = 1, nivel_n_suportado = 2, file_not_found = 3
};
int tabela_quantidade_caracteres[][3] = {{10, 12, 14},
                                         {9,  11, 13},
                                         {8,  16, 16},
                                         {8,  10, 12}};

string err2str(int code);

void QR::analise() {
    modo_codificacao = NUMERICO;
    for (const char *c = msg.c_str(); *c; c += 1) {
        if (isdigit(*c))
            continue;
        if (isupper(*c) || *c == ' ' || *c == '$' || *c == '*' || *c == '+' || *c == '.' || *c == '/' ||
            *c == ':') {
            modo_codificacao = ALPHANUMERICO;
            continue;
        }
        modo_codificacao = BYTE;
        break;
    }
}

void QR::codificar() {
    if (!this->tentar_todos_modos_de_correcao) {
        this->modo_correcao = H;
    }
    while (this->modo_correcao >= L) {
        this->error = 0;
        determinarMenorVersao();
        indicadorModo();
        contagemCaracteres();
        codificarDados();
//     this->error dividir_em_blocos();
        if (this->error) {
            if (!this->tentar_todos_modos_de_correcao && this->modo_correcao > L) {
                this->modo_correcao--;
                continue;
            }
            this->modo_correcao = -1;

        }
        return;
    }

}

void QR::determinarMenorVersao() {
    FILE *tabela = nullptr;
    int caracteres[4];
    switch (modo_correcao) {
        case H:
            tabela = fopen("../src/tabelas/versoes/H.txt", "r");
            break;
        case Q:
            tabela = fopen("../srg/tabelas/versoes/Q.txt", "r");
            break;
        case M:
            tabela = fopen("../src/tabelas/versoes/M.txt", "r");
            break;
        case L:
            tabela = fopen("../src/tabelas/versoes/L.txt", "r");
            break;
    }
    if (!tabela) {
        error = file_not_found;
        return;
    }
    for (int i = 0; i < 40 && !feof(tabela); ++i) {
        fscanf(tabela, "%d %d %d %d", caracteres, caracteres + 1, caracteres + 2, caracteres + 3);
        if (tamanhoMensagem <= caracteres[modo_codificacao - 1]) {
            capacidadeCaracteres = caracteres[modo_codificacao - 1];
            versao = i + 1;
            fclose(tabela);
            return;
        }
    }
    fclose(tabela);
    error = nivel_n_suportado;
}

string dec2bin(int number, int bits) {
    string ans = "";
    for (; bits > 0; bits--) {
        ans = to_string(number % 2) + ans;
        number = number / 2;
    }
    return ans;
}

void QR::indicadorModo() {
    strbits = dec2bin(pow(2, modo_codificacao - 1), 4);
}

void QR::contagemCaracteres() {
    int bits = 0;
    if (versao >= 27)
        bits = 2;
    else if (versao >= 10)
        bits = 1;
    bits = tabela_quantidade_caracteres[modo_codificacao - 1][bits];
    strbits += dec2bin(tamanhoMensagem, bits);
}

int convert(char c) {
    if (c <= '9' && c >= '0')
        return c - '0';
    if (c <= 'Z' && c >= 'A')
        return c - 'A' + 10;
    if (c == ' ')return 36;
    else if (c == '$')
        return 37;
    else if (c == '%')
        return 38;
    else if (c == '*')
        return 39;
    else if (c == '+')
        return 40;
    else if (c == '-')
        return 41;
    else if (c == '.')
        return 42;
    else if (c == '/')
        return 43;
    else if (c == ':')
        return 44;
}

void QR::codificarDados() {
    string part, bloco;
    int number;
    if (modo_codificacao == NUMERICO) {
        for (int i = 0; i < tamanhoMensagem; i += 3) {
            part = msg.substr(i, i + 3);
            number = stoi(part);
            bloco = dec2bin(number, number < 10 ? 4 : number < 100 ? 7 : 10);
            strbits += bloco;
        }
    } else if (modo_codificacao == ALPHANUMERICO) {
        for (int i = 0; i < tamanhoMensagem; i += 2) {
            part = msg.substr(i, i + 2);
            if (part.length() == 1) {
                number = convert(part[0]);
                bloco = dec2bin(number, 6);
            } else {
                number = convert(part[0]) * 45 + convert(part[1]);
                bloco = dec2bin(number, 11);
            }
            strbits += bloco;
        }
    } else if (modo_codificacao == BYTE) {
        for (int i = 0; i < tamanhoMensagem; i ++) {
            bloco = dec2bin(msg[i], 8);
            cout<<msg[i]<<' '<<bloco<<endl;
            strbits += bloco;
        }
    }
}

std::ostream &operator<<(std::ostream &os, QR &v) {
    if (v.error) {
        os << "erro: " << err2str(v.error);
        return os;
    }
    os << "msg: " << v.msg <<
       "\nmodo de codificacao: " << v.modo_codificacao <<
       "\nModo de correcao: " << v.modo_correcao <<
       "\nVersao: " << v.versao <<
       "\nTamanho da msg: " << v.tamanhoMensagem <<
       "\ncapacidade de caracteres: " << v.capacidadeCaracteres <<
       "\nStr bits : " << v.strbits;
    return os;
}

string err2str(int code) {
    switch (code) {
        case Msg_n_suport:
            return "mensagem nao suportada";
        case nivel_n_suportado:
            return "nivel_n_suportado nao suportada";
        case file_not_found:
            return "Arquivo nao encontrado";
        case 0:
            return "";
    }
    return "falha na geracao do qr";
}

File: QR-code/source code/python/src/QR.py
NUMERICO = 1
ALPHANUMERICO = 2
BYTE = 3
L = 1
M = 2
Q = 3
H = 4
tabela = {'quantidade de caracteres': [[10, 12, 14], [9, 11, 13], [8, 16, 16], [8, 10, 12]],}


class QR_exception(Exception):
    def __init__(self, *args: object) -> None:
        super().__init__(*args)


def dec2bin(number, bits):
    ans = ''
    for i in range(bits):
        ans = str(number % 2) + ans
        number = number // 2
    return ans


class QR:
    def __init__(self, msg, forcar_modo_de_correcao=False):
        self.capacidadeCaracteres = None
        self.msg = msg
        self.tamanho_da_mensagem = len(msg)
        self.modo_correcao_forcado = forcar_modo_de_correcao
        self.modo_correcao = H
        self.versao = None
        self.strbits = None
        # etapa 1-7 analise de dados

    def __repr__(self):
        s = ''
        for i, v in self.__dict__.items():
            s += f'{str(i).replace("_", " ")} : {v}\n'
        return s

    def analise(self):
        self.modo_codificao = NUMERICO
        for c in self.msg:
            if c.isdigit():
                continue
            if 'A' <= c <= 'Z' or c in [' ', '$', '*', '+', '.', '/']:
                self.modo_codificao = ALPHANUMERICO
                continue
            self.modo_codificao = BYTE
            break

    def codificar(self, flag=None):
        if not self.modo_correcao_forcado and flag != 'repeat':
            self.modo_correcao = H
        try:
            self.determinarMenorVersao()
            self.indicadorModo()
            self.contagemCaracteres()
            self.codificarDados()
            # self.dividir_em_blocos()
        except QR_exception as e:
            if self.modo_correcao_forcado and self.modo_correcao > L:
                self.modo_correcao -= 1
                self.codificar('repeat')
                return
            raise QR_exception("mensagem nao suportada")

    def determinarMenorVersao(self):
        titulo = ['L', 'M', 'Q', 'H']
        tabela = open(f"tabelas/versoes/{titulo[self.modo_correcao - 1]}.txt", "r", encoding='utf-8')
        for i in range(40):
            caracteres = [int(j) for j in tabela.readline().split(' ')]
            if self.tamanho_da_mensagem <= caracteres[self.modo_codificao - 1]:
                self.capacidadeCaracteres = caracteres[self.modo_codificao - 1]
                self.versao = i + 1
                tabela.close()
                return
        tabela.close()
        raise QR_exception('modo nao suportado')

    def indicadorModo(self):
        self.strbits = dec2bin(2 ** (self.modo_codificao - 1), 4)

    def contagemCaracteres(self):
        if self.versao >= 27:
            bits = 2
        elif self.versao >= 10:
            bits = 1
        else:
            bits = 0
        bits = tabela['quantidade de caracteres'][self.modo_codificao][bits]
        self.strbits = self.strbits + dec2bin(self.tamanho_da_mensagem, bits)

    def codificarDados(self):
        if self.modo_codificao == NUMERICO:
            self.codificarNumerico()
        elif self.modo_codificao == ALPHANUMERICO:
            self.codificarAlphaumerico()
        elif self.modo_codificao == BYTE:
            self.codificarByte()

    def codificarNumerico(self):
        i = 0
        while i < self.tamanho_da_mensagem:
            number = int(self.msg[i:i + 3])
            bloco = dec2bin(number, number < 10 and 4 or number < 100 and 7 or 10)
            self.strbits += bloco
            i += 3

    def codificarAlphaumerico(self):
        def converter(c):
            if '0' < c < '9':
                return int(c)
            if 'A' < c < 'Z':
                return ord(c) - ord('A') + 10
            return {' ': 36, '$': 37, '%': 38, '*': 39, '+': 40, '-': 41, '.': 42, '/': 43, ':': 44}[c]

        i = 0
        while i < self.tamanho_da_mensagem:
            dupla = self.msg[i:i + 2]
            print(f'"{dupla}"',end=' -> ')
            if len(dupla) == 1:
                number = converter(dupla)
                bloco = dec2bin(number, 6)
            else:
                number = converter(dupla[0]) * 45 + converter(dupla[1])
                bloco = dec2bin(number, 11)
            print(bloco)
            self.strbits += bloco
            i += 2

    def codificarByte(self):
        for c in self.msg:
            bloco = dec2bin(ord(c), 8)
            self.strbits += bloco
